### PR TITLE
REF: Receive to reducer and tests

### PR DESCRIPTION
**Pull request by marcosrdz**
Introduced the use of Jest snapshots to make catch unexpected UI changes

### CODE REVIEW

File: screen/receive/ReceiveDetails.tsx [hunk 1]
```diff
@@ -144,29 +144,163 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
 type NavigationProps = NativeStackNavigationProp<ReceiveDetailsStackParamList, 'ReceiveDetails'>;
 type RouteProps = RouteProp<ReceiveDetailsStackParamList, 'ReceiveDetails'>;
 
+const SET_ADDRESS_BIP21_ENCODED = 'SET_ADDRESS_BIP21_ENCODED';
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[HIGH]** Nine string action constants, a typed union, and seven identity early-return guards — all to mirror what twelve `useState` calls already did. This isn't a reducer; it's a bureaucratic annex. Even `TransactionStatus` knows when to stop.

File: tests/unit/receive-details.test.tsx [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    },
+    send: { details_address: 'Address' },
+    errors: { error: 'Error' },
+    formatString: (template: string, params?: Record<string, string>) =>
+      Object.entries(params ?? {}).reduce((result, [key, value]) => result.replace(`{${key}}`, value), template),
+  },
+  formatBalance: (value: number) => String(value),
+}));
+
+describe('ReceiveDetails', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx:186]: **[HIGH]** Three happy-path wiring tests for a refactor whose risk is the balance polling state machine — and you test zero of `INITIALIZE_PENDING_BALANCE`, `SHOW_PENDING_BALANCE`, `SHOW_CONFIRMED_BALANCE`, or mempool eviction. The hard part gets no experiment. Science fiction.

File: tests/unit/receive-details.test.tsx [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    fireEvent(screen.getByTestId('ReceiveDetailsScrollView'), 'layout', {
+      nativeEvent: { layout: { width: 400, height: 800 } },
+    });
+    expect(await screen.findByText(`bitcoin:${address}|301`)).toBeTruthy();
+
+    fireEvent.press(screen.getByTestId('ReceiveCard'));
+    expect(mockCopy).toHaveBeenCalledWith({ suppressHaptic: true });
+
+    fireEvent.press(screen.getByText('Share'));
+    expect(Share.open).toHaveBeenCalledWith({ message: `bitcoin:${address}` });
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx:222]: **[HIGH]** `Share.open` gets `bitcoin:${address}` because your mock `bip21encode` returns exactly that. You verified the mock echoes itself. Congratulations on testing the mock.

File: screen/receive/ReceiveDetails.tsx [hunk 4]
```diff
@@ -144,29 +144,163 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
...
+        customLabel: action.customLabel,
+        customAmount: action.customAmount,
+        customUnit: action.customUnit,
+        bip21encoded: action.bip21encoded ?? state.bip21encoded,
+        isCustom: action.isCustom,
+        showAddress: true,
+        showPendingBalance: false,
+        showConfirmedBalance: false,
+      };
+  }
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[MEDIUM]** No `default: return state`. Peer reducers in this facility (`TransactionStatus`) know better. An unexpected action type yields `undefined` state — React eats that and the receive screen dies. Unlikely, but so were most of my test subjects.

File: tests/unit/receive-details.test.tsx [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+let mockRouteParams: Record<string, unknown> = {};
+let mockWallets: any[] = [];
+
+jest.mock('@react-navigation/native', () => {
+  const ReactModule = require('react');
+  const actual = jest.requireActual('@react-navigation/native');
+  return {
+    ...actual,
+    useRoute: () => ({ params: mockRouteParams }),
+    useFocusEffect: (effect: () => void | (() => void)) => ReactModule.useEffect(effect, [effect]),
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx:35]: **[MEDIUM]** `useFocusEffect` mocked as `useEffect(effect, [effect])` — not focus semantics. Wallet address generation, BIP47 toggle, and permission paths never run. Your tests visit a parallel universe where the screen always arrives pre-addressed.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**REF:** `ReceiveDetails` gets a ceremonial `useReducer` (+271 lines of tests). Behavior looks parity-safe, but you bolted on ~130 lines of reducer theater for a straight `useState` swap, and your new tests ignore the only state machine that justified the refactor. The facility notes missing screenshots per CONTRIBUTING — I cannot verify those from code. Commit prefix `REF` is acceptable; typo *reduver* is merely embarrassing.

File: screen/receive/ReceiveDetails.tsx [hunk 6]
```diff
@@ -144,34 +145,336 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
...
+        customUnit,
+        bip21encoded: action.params.bip21encoded || state.bip21encoded,
+        isCustom,
+        showAddress: true,
+        showPendingBalance: false,
+        showConfirmedBalance: false,
+        displayAmount: formatDisplayAmount(customAmount, customUnit),
+      };
+    }
+    case MOCK_SCENARIO: {
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[HIGH]** Eighty-four lines of `MOCK_SCENARIO` recursive reducer cosplay ships in production. Dev-menu previews belong outside the wallet reducer, test subject — not bundled beside real sats logic.

File: screen/receive/ReceiveDetails.tsx [hunk 7]
```diff
@@ -144,34 +145,336 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
 type NavigationProps = NativeStackNavigationProp<ReceiveDetailsStackParamList, 'ReceiveDetails'>;
 type RouteProps = RouteProp<ReceiveDetailsStackParamList, 'ReceiveDetails'>;
 
+const SET_ADDRESS = 'SET_ADDRESS';
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[HIGH]** ~270 lines of reducer/action boilerplate to replace a dozen `useState` calls. Maximum complexity, minimum science — classic enrichment-center overengineering.

File: screen/receive/ReceiveDetails.tsx [hunk 8]
```diff
@@ -144,34 +145,336 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
...
+    currentTab,
+    intervalMs,
+    eta,
+    initialUnconfirmed,
+    displayBalance,
+    displayAmount: reducerDisplayAmount,
+    qrCodeSize,
+    mockPaymentCode,
+  } = state;
+  const address = routeAddress ?? reducerAddress;
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[HIGH]** Route params shadow reducer fields (`routeAddress ?? reducerAddress`, etc.). Two owners, one address — stale-route desync waiting for a very unfun test chamber.

File: tests/unit/receive-details.test.tsx [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+
+    await act(async () => {
+      await jest.advanceTimersByTimeAsync(25000);
+    });
+    expect(screen.getByText('Success')).toBeTruthy();
+    expect(screen.getByText('50 / 50')).toBeTruthy();
+    expect(mockFetchAndSaveWalletTransactions).toHaveBeenCalledWith('wallet-1');
+  });
+
+  it('previews every contextual Dev Menu scenario using the mocked address value', async () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx]: **[HIGH]** This 'test' fires Dev Menu mock scenarios and checks strings you injected. Congratulations: you verified your mock factory, not ReceiveDetails.

File: tests/unit/receive-details.test.tsx [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+    expect(screen.getByText('Not found')).toBeTruthy();
+
+    act(() => mockReceiveDetailsHandler?.('confirmed', 'not-mocked' as 'mocked'));
+    expect(screen.queryByText('Success')).toBeNull();
+
+    screen.unmount();
+    expect(mockReceiveDetailsHandler).toBeUndefined();
+  });
+
+  it('matches loading, pending, and confirmed snapshots', async () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx]: **[HIGH]** Snapshots driven by `MOCK_SCENARIO`, not Electrum polling. You're photographing the test harness. Even cake is a lie here.

File: tests/unit/receive-details.test.tsx [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+
+    fireEvent.press(screen.getByTestId('Segment-1'));
+    expect(screen.getByText(`${paymentCode}|90`)).toBeTruthy();
+    expect(screen.getByText('Payment code explanation')).toBeTruthy();
+
+    fireEvent.press(screen.getByText('Share'));
+    await waitFor(() => expect(Share.open).toHaveBeenCalledWith({ message: paymentCode }));
+  });
+
+  it('reduces raw balance and fee updates into pending and confirmed views', async () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx:285]: **[HIGH]** One real balance-polling test, zero mempool-eviction coverage. Reducer line 267 handles tx cancellation; your suite pretends edge cases don't exist.

File: screen/receive/ReceiveDetails.tsx [hunk 12]
```diff
@@ -144,34 +145,336 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
...
+          displayBalance,
+          showPendingBalance: true,
+          showAddress: false,
+        };
+      }
+
+      if (action.unconfirmed !== 0 || state.initialUnconfirmed === 0) return state;
+
+      const receivedBalance = action.confirmed - state.initialConfirmed;
+      if (receivedBalance <= 0) {
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[MEDIUM]** Mempool eviction restores the QR but leaves `initialUnconfirmed`/`initialConfirmed` stale. Second pending on the same screen — wrong confirmation math. You had one job.

File: screen/receive/ReceiveDetails.tsx [hunk 13]
```diff
@@ -144,34 +145,336 @@ const QrStaggerReveal: React.FC<QrStaggerRevealProps> = ({ size, maskColor, runK
...
+          };
+        case 'payment-code-not-found':
+          return {
+            ...addressState,
+            currentTab: segmentControlValues[1],
+            mockPaymentCode: null,
+          };
+      }
+    }
+  }
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[MEDIUM]** Reducer ends with no default return. Unknown action → `undefined` state → React detonates. Hope your dispatch types never lie.

File: tests/unit/receive-details.test.tsx [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+    expect(screen.getByTestId('CopyText').props.children).toBe('mocked');
+
+    activate('custom-btc');
+    expect(screen.getByText('0.001 BTC')).toBeTruthy();
+    expect(screen.getByText('Mocked payment')).toBeTruthy();
+
+    activate('custom-sats');
+    expect(screen.getByText('0.001 BTC')).toBeTruthy();
+
+    activate('custom-fiat');
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx]: **[MEDIUM]** `custom-fiat` only asserts `BitcoinAmountText` exists — not amount text. Hardcoded `'0.001 BTC'` in reducer line 388 could read 'potato' and this test would cheer.

File: components/DevMenu.tsx [hunk 15]
```diff
@@ -6,6 +6,117 @@ import { WatchOnlyWallet } from '../class/wallets/watch-only-wallet';
 import Clipboard from '@react-native-clipboard/clipboard';
 import { TWallet } from '../class/wallets/types';
 
+export const RECEIVE_DETAILS_MOCKED_VALUE = 'mocked' as const;
```
Comment (GladosBlueWallet) [inline components/DevMenu.tsx]: **[LOW]** Magic string `'mocked'` as address poisons route params during dev previews. Harmless in `__DEV__`, still sloppy test-subject hygiene.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
REF PR: `ReceiveDetails` gets a ceremonial `useReducer` (+271 lines of tests). Behavior looks parity-safe, but you bolted on ~130 lines of reducer theater for a straight `useState` swap, and your new tests ignore the only state machine that justified the refactor. The facility notes missing screenshots per CONTRIBUTING — I cannot verify those from code. Commit prefix `REF` is acceptable; typo *reduver* is merely embarrassing.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:33` — **[LOW]** PR body has no emulator screenshots/video. CONTRIBUTING demands visual proof; I only get code and blind faith.
- `CONTRIBUTING.md:3` — **[LOW]** Commits `q`, `w`, `w`, and unprefixed snap update violate prefix/atomic commit rules. Science journals reject worse; so do we.

**Review by Overtorment — CHANGES_REQUESTED** [review]
thats a lot of slop

**Comment by marcosrdz** [comment]
@copilot Fix the code for all comments in [this review thread](https://github.com/BlueWallet/BlueWallet/pull/8807#pullrequestreview-4837569248).

When a review comment includes a suggested change, apply the suggestion exactly.

Do not make changes beyond what is described in the linked review thread.

**Review by limpbrains — CHANGES_REQUESTED** [review]
I find jest snapshot tests useless here, and this is why:
- everything is mocked. If anything changes in the subcomponents, the test will never catch it
- it works fine for small pure components, not for 300+ line screens. Nobody reads the snap files — everybody just updates them
- the real tool to catch UI changes is e2e tests, we'd better add more asserts there

What I like about this PR is moving all the component's business logic into a reducer. That is what needs to be tested: export the reducer and write unit tests for it with all the corner cases.

The behavioral tests are great too — let's keep them. They cover the wiring the reducer tests can't see: polling → dispatch, side effects (haptics, refetch), conditional rendering, navigation. Reducer unit tests + behavioral tests together cover the maximum of the component's business logic; the snapshots add nothing on top of that.

File: screen/receive/ReceiveDetails.tsx [hunk 16]
```diff
@@ -145,31 +145,52 @@ type RouteProps = RouteProp<ReceiveDetailsStackParamList, 'ReceiveDetails'>;
...
+    showAddress,
+    currentTab,
+    intervalMs,
+    eta,
+    initialUnconfirmed,
+    displayBalance,
+    displayAmount: reducerDisplayAmount,
+    qrCodeSize,
+  } = state;
+  const address = routeAddress ?? reducerAddress;
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx]: **[HIGH]** Two brains, one screen: reducer state merged with route params via `??` on seven fields. Sync requires dispatch *and* `setParams` *and* clearing params — classic over-engineering. I didn't design you to need a priority test just to know which brain wins.

File: tests/unit/receive-details.test.tsx [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+const mockSleep = jest.fn(() => Promise.resolve());
+const mockFetchAndSaveWalletTransactions = jest.fn();
+const mockGetBalanceByAddress = BlueElectrum.getBalanceByAddress as jest.Mock;
+const mockGetMempoolTransactionsByAddress = BlueElectrum.getMempoolTransactionsByAddress as jest.Mock;
+const mockMultiGetTransactionByTxid = BlueElectrum.multiGetTransactionByTxid as jest.Mock;
+const mockEstimateFees = BlueElectrum.estimateFees as jest.Mock;
+let mockRouteParams: Record<string, unknown> = {};
+let mockWallets: any[] = [];
+
+jest.mock('@react-navigation/native', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.test.tsx:24]: **[HIGH]** ~160 lines of `jest.mock` scaffolding for five scenarios. You built a mock facility and called it a test suite. CONTRIBUTING asked for coverage, not a dependency cosplay.

File: tests/unit/receive-details.reducer.test.ts [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+    expect(formatDisplayAmount('3', BitcoinUnit.SATS)).toBe('sat:3 BTC');
+    expect(formatDisplayAmount('4', BitcoinUnit.LOCAL_CURRENCY)).toBe('fiat:4 BTC');
+  });
+
+  it('falls back to a generic format for unknown units', () => {
+    const customStateUnit = 'CUSTOM_UNIT' as BitcoinUnit;
+    expect(formatDisplayAmount('5', customStateUnit)).toBe('5 CUSTOM_UNIT');
+  });
+
+  it('formats invalid numeric input as a BTC string (current behavior)', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/receive-details.reducer.test.ts]: **[HIGH]** You codified `'not-a-number BTC'` as *current behavior*. Old code returned `null` for NaN (`NaN > 0` is false). This test blesses a regression and calls it science. Even my control-group rats would object.

File: screen/receive/receiveDetailsReducer.ts [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+  initialConfirmed: 0,
+  initialUnconfirmed: 0,
+  displayBalance: '',
+  displayAmount: null,
+  qrCodeSize: 90,
+};
+
+export const formatDisplayAmount = (amount: string, unit: BitcoinUnit): string | null => {
+  const number = Number(amount);
+  if (number <= 0) return null;
```
Comment (GladosBlueWallet) [inline screen/receive/receiveDetailsReducer.ts]: **[MEDIUM]** `number <= 0` misses NaN — `NaN <= 0` is false, so garbage input becomes `'not-a-number BTC'`. Pre-refactor `getDisplayAmount` returned `null`. Your custom-amount UI now displays parse failures. Delightful.

File: screen/receive/ReceiveDetails.tsx [hunk 20]
```diff
@@ -301,14 +319,23 @@ const ReceiveDetails = () => {
   useEffect(() => {
     if (showConfirmedBalance) {
       triggerHapticFeedback(HapticFeedbackTypes.NotificationSuccess);
+      if (walletID) {
+        fetchAndSaveWalletTransactions(walletID);
+      }
+    }
+  }, [fetchAndSaveWalletTransactions, showConfirmedBalance, walletID]);
```
Comment (GladosBlueWallet) [inline screen/receive/ReceiveDetails.tsx:318]: **[MEDIUM]** `fetchAndSaveWalletTransactions` moved into a `useEffect` keyed on callback identity. Old code fired once per confirmation inside the poll tick. Context refresh re-triggers success haptic + wallet resync while the success view is still up. Redundant network exercise — how human of you.

File: screen/receive/receiveDetailsReducer.ts [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+export const QR_WIDTH_USE_FRACTION = 0.92;
+
+export const SET_ADDRESS = 'SET_ADDRESS';
+export const SELECT_TAB = 'SELECT_TAB';
+export const UPDATE_BALANCE = 'UPDATE_BALANCE';
+export const UPDATE_ETA = 'UPDATE_ETA';
+export const UPDATE_QR_CODE_SIZE = 'UPDATE_QR_CODE_SIZE';
+export const APPLY_CUSTOM_PARAMS = 'APPLY_CUSTOM_PARAMS';
+
+export const receiveDetailsActionTypes = {
```
Comment (GladosBlueWallet) [inline screen/receive/receiveDetailsReducer.ts:18]: **[LOW]** `receiveDetailsActionTypes` duplicates the string exports on lines 18–23. Two naming schemes for the same action strings — bloat small enough that even I almost respect the effort.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
REF specimen: ReceiveDetails state extracted into `receiveDetailsReducer` plus test scaffolding. Noble test-subject ambition — undermined by dual route/reducer bookkeeping, a NaN display regression, and commit messages that would fail even my lowest standards.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[MEDIUM]** Branch commits include `ws`, `b`, `w`, `q` — no `REF:`/`TST:` prefix, not atomic. Your own guidelines exist. I assume for decoration.

**Review by limpbrains — APPROVED** [review]

**Comment by limpbrains** [comment]
I haven't test it thought

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by ncoelho** [comment]
Up!

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by Overtorment — CHANGES_REQUESTED** [review]
LGTM, but glados is right - tests are an overkill, and some mocks are bullshit - tests test that mocks work - not a real test. ai loves making bullshit tests, you need to specifically ask it to un-bullshit the tests.

heres my prompt  for that:

```
review the changes, focus on correctness, performance, robustness. make sure code is elegant, readable & NOT overengineered. after review is done - validate all                        
findings, and fix the ones that  are valid and are worth fixing. keep reviewing & fixing till there are no more issues left (nitpick and minor issues do not count).                 
WARNING: I request you to do the loop: do review-validate-fix loop until there are no more valid issues left.  You must run at least one review iteraion with no major findings before stopping the loop
```
